### ISSUE TITLE
buffer metadata and expose info to templates before compile

### ISSUE THREAD

**Issue by declandewet** [carrot/roots-dynamic-content#22]
I'm working on a project for a client, which involves categorised articles - each article needs to display 3 random articles from the same category underneath it's rendered content. I've found this is not possible with this extension, and think buffering the dynamic content info in-memory before each content file is rendered to HTML will help mitigate the issue.

As a workaround to this issue, I am loading the content in via AJAX and the JSON export feature, however this is not ideal as the client has requested this site to be functional on feature phones that do not support JavaScript.

For the purpose of this issue, I feel the need to exhibit a general idea of the directory structure for these articles:

```
./articles/
----mama-to-mama/
--------month-1.jade
--------month-2.jade
----baby/...
----pregnancy/...
```

I've also created a few mockups that should help to better illustrate the problem:

This is the expected output/wanted functionality for a single article:
![Imgur](http://i.imgur.com/XAqiqZp.png)

This is the *actual* output for each article:
![Imgur](http://i.imgur.com/M1MkcRo.png)
![Imgur](http://i.imgur.com/4OLpzwx.png)
![Imgur](http://i.imgur.com/IPJmDV6.png)

As you can see, since each file is rendered one after the other, the information is not buffered in time for the rendered HTML to display all of the content.

**Comment by jenius** [comment]
Yeah so this is an issue that has come up often, and the export json feature was made specifically to deal with it. The reason this happens is because roots runs through dynamic content one by one. So from any given piece of dynamic content, unless it is the last one compiled, it will not have access to all other dynamic content, fully rendered. This is not a bug, it's a limitation of static sites.

The only way to solve this otherwise is to run a full compile twice. The first time, to get all the content compiled, and the second time to compile again where each post has access to all other compiled posts. Not only is this a large waste of time and resources, but also it's not really feasible from an extension as this is far too invasive of an override to be made to the way roots works.

That being said, I feel like it's fairly unlikely that this functionality will be made with the information I currently know. If anyone else has a solution that would work better I'd love to hear about it!

**Comment by declandewet** [comment]
Is it not possible for Roots to hold off on compiling these files until it has read all of the source files and stored their content in-memory, then write the output files once only all the source file information has been stored? Seems like a simple thing to do in Node. Otherwise, there are non-javascript alternatives, such as creating a page for the "other blog posts" for each category, and loading that in using either an `<iframe>` or `<object data=...>` - but that's a bit hacky/dirty/bad for seo - any thoughts?

**Comment by jenius** [comment]
The problem is that it doesn't just read the files, it does parsing and compiling as well before making the metadata available. For a blog post, if you are using jade, dynamic content allows the contents to be available, and uses jade to compile that.

**Comment by declandewet** [comment]
Do you think it might be possible to read in the front matter first and at least make *that* available before the files are rendered out? It seems like for this kind of use case, having the content available is not really needed - but having the front matter available would definitely be handy so one could render out titles/dates/author/descriptions/etc

**Comment by jenius** [comment]
Maybe, I'd have to dig into it pretty deeply, and unfortunately I don't really have the time to do so right now :cry:

**Comment by terusus** [comment]
Hi.

I ran into a similar problem. I was trying to access dynamic content from other dynamic content and this did not go well for me. Further more, because of that fact that the front matter and content are parsed/rendered together, experimenting with such access causes different results on windows and linux.

It would be a huge advantage to parse the front matter in advance and have the _content variable as a promise, for example.

I am willing to try and implement this but I do not understand the code very well. However, with some explanation I can get to work.

**Comment by argyleink** [comment]
i'm willing to help resolve this too, if we could get a quick breakdown of where to look in the code..

**Comment by andyshora** [comment]
@jenius @argyleink @declandewet this is one of the best bug descriptions I've seen! I've just hit this problem too. Could anyone please recommend a solution?

**Comment by zchsh** [comment]
I'm kinda new to roots, and I just want simple < previous next > links on each post, so having all of the front matter available to render would be really useful to me as well. Right now, I'm using roots-records to kind of bridge the gap: I set the option to get dynamic-content to export JSON, then move that JSON output into the roots-records folder, and then I can access all the front matter (from the previous compile, obvs) using roots-records. I don't know if that really qualifies as a 'solution' though 😬

**Comment by terusus** [comment]
I ended up using [contentful](https://www.contentful.com) and render that data with the static generator. If you think about it - it makes more sense. It will be possible to scale in the future. It's easier for the content writers. And the free tier is more than enough for a small site.
